Do not wrap output with input device inside _to_copy

laithsakka opened the pull request:
Fixing https://github.com/pytorch/pytorch/issues/118790

This diff revert a small part of the code that was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
but its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

When such conversion does not happen we get the following error
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to
 a python object of type FakeTensor
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119868



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

laithsakka committed:
work

[ghstack-poisoned]

laithsakka committed:
Update on "work"




[ghstack-poisoned]

laithsakka committed:
Update on "Testing removing a recent change to see failures if any"

testing a theory for fixing https://github.com/pytorch/pytorch/issues/118790

The code reverted in this diff was introduced in https://github.com/pytorch/pytorch/pull/104689

I think maybe a proper solution in case there is a risk that a fake tensot input has become a non fake  is to 
check the following:

if the input was originally fake tensor and output has become non fake, then we wrap output with input device?





[ghstack-poisoned]

laithsakka commented:
@VRSinghHabana

laithsakka committed:
Update on "Assert that dtype promotion does not convert fake tensor to tensor and remove bad double fake tensor wrapping"

testing a theory for fixing https://github.com/pytorch/pytorch/issues/118790

The code reverted in this diff was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above claims added a comment that claims "In case of dtype promotion, faketensor converted into tensor"
the path above results in trying to wrap a fake tensor with a fake tensor resulting and an error
```
torch._dynamo.exc.TorchRuntimeError: Failed running call_method to(*(FakeTensor(..., size=(2,)), 'meta'), **{}):
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor
```

I debugged and did not see that type promotion is changing a fake tensor to a tensor, I removed that code
and everything pass well with no failures, hence I added an assertion that if the input is fake tensor then type 
promotion is keeping it fake and removed that added call. 
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor
```

if the input was originally fake tensor and output has become non fake, then we wrap output with input device?





[ghstack-poisoned]

laithsakka committed:
Update on "Only wrap_output_with_input_device_ if dtype conversion changed a fake tensor to tensor."

testing a theory for fixing https://github.com/pytorch/pytorch/issues/118790

The code reverted in this diff was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

I debugged and did not saw that when such conversion does not happen the following error happen. 
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor
```

I updated the code to only call wrap_output_with_input_device_ if the input is fake and became no fake after the conversion.
I wonder though if the dtype conversion itself should preserve the fakeness of the tensor instead.
and if we shall update _convert_element_type_meta(a: TensorLikeType, dtype: torch.dtype) instead




[ghstack-poisoned]

laithsakka committed:
Update on "Only wrap_output_with_input_device_ if dtype conversion changed a fake tensor to tensor."

testing a theory for fixing https://github.com/pytorch/pytorch/issues/118790

The code reverted in this diff was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

I debugged and did not saw that when such conversion does not happen the following error happen. 
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor
```

I updated the code to only call wrap_output_with_input_device_ if the input is fake and became no fake after the conversion.
I wonder though if the dtype conversion itself should preserve the fakeness of the tensor instead.
and if we shall update _convert_element_type_meta(a: TensorLikeType, dtype: torch.dtype) instead




[ghstack-poisoned]

ezyang commented on torch/_decomp/decompositions.py:
```diff
@@ -1891,19 +1890,25 @@ def _to_copy(
         return x.clone()
     dtype_converted = False
     common_device = device_hint(x)
+    inputIsFake = isinstance(x, torch._subclasses.FakeTensor)
```
nit: conventionally locals in Python use snake case, rather than camel case. The naming here is inconsistent even with the variables above and it sticks out like a sore thumb. Stick to local conventions.

ezyang commented:
I'm not entirely convinced by the approach here, but I'm also not sure what I'd counterpropose. If you want to hop on chat to talk about it I'm down.

laithsakka commented:
> I'm not entirely convinced by the approach here, but I'm also not sure what I'd counterpropose. If you want to hop on chat to talk about it I'm down.

yes indeed.

laithsakka committed:
Update on "Only wrap_output_with_input_device_ if dtype conversion changed a fake tensor to tensor."

testing a theory for fixing https://github.com/pytorch/pytorch/issues/118790

The code reverted in this diff was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
but its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

When such conversion does not happen we get the following error
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to
 a python object of type FakeTensor
```

I updated the code to only call wrap_output_with_input_device_ if the input is fake and became no fake after the conversion. I wonder though if the dtype conversion itself should preserve the fakeness of the tensor instead.
and if we shall update _convert_element_type_meta(a: TensorLikeType, dtype: torch.dtype) instead




[ghstack-poisoned]

laithsakka committed:
Update on "Only wrap_output_with_input_device_ if dtype conversion changed a fake tensor to tensor."

Fixing https://github.com/pytorch/pytorch/issues/118790

The code reverted in this diff was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
but its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

When such conversion does not happen we get the following error
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to
 a python object of type FakeTensor
```

I updated the code to only call wrap_output_with_input_device_ if the input is fake and became no fake after the conversion. I wonder though if the dtype conversion itself should preserve the fakeness of the tensor instead.
and if we shall update _convert_element_type_meta(a: TensorLikeType, dtype: torch.dtype) instead




[ghstack-poisoned]

laithsakka committed:
Update on "Only wrap_output_with_input_device_ if dtype conversion changed a fake tensor to tensor."

Fixing https://github.com/pytorch/pytorch/issues/118790

The code reverted in this diff was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
but its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

When such conversion does not happen we get the following error
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to
 a python object of type FakeTensor
```

I updated the code to only call wrap_output_with_input_device_ if the input is fake and became no fake after the conversion. I wonder though if the dtype conversion itself should preserve the fakeness of the tensor instead.
and if we shall update _convert_element_type_meta(a: TensorLikeType, dtype: torch.dtype) instead




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
The new agreement is we're just going to yank the Habana patch.

laithsakka committed:
Update on "Do not wrap output with input device inside _to_copy"

Fixing https://github.com/pytorch/pytorch/issues/118790

This diff revert a small part of the code that was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
but its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

When such conversion does not happen we get the following error
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to
 a python object of type FakeTensor
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

laithsakka commented:
rebase

thiagocrepaldi reviewed (approved):
Thanks. This also fixes #120323

thiagocrepaldi commented:
Hi @laithsakka thanks for this PR!

Looking the CI results, it seems the PR has 1) lint issues that can be fixed by `lintrunner -a` and a few occurences of 2) `FAILED [0.0526s] test_fake_tensor.py::FakeTensorConverterTest::test_memoized_conversion_from_meta - RuntimeError: Unexpected success, please remove test from dynamo_test_failures.py` that seems to be minor too

laithsakka committed:
Update on "Do not wrap output with input device inside _to_copy"

Fixing https://github.com/pytorch/pytorch/issues/118790

This diff revert a small part of the code that was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
but its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

When such conversion does not happen we get the following error
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to
 a python object of type FakeTensor
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

laithsakka committed:
Update on "Do not wrap output with input device inside _to_copy"

Fixing https://github.com/pytorch/pytorch/issues/118790

This diff revert a small part of the code that was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
but its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

When such conversion does not happen we get the following error
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to
 a python object of type FakeTensor
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

thiagocrepaldi commented:
`2024-02-26T21:48:17.9111397Z FAILED [0.0878s] test_serialization.py::TestSubclassSerialization::test_tensor_subclass_getstate_overwrite - RuntimeError: Unexpected success, please remove test from dynamo_test_failures.py`

laithsakka commented:
@pytorchbot merge

laithsakka committed:
Update on "Do not wrap output with input device inside _to_copy"

Fixing https://github.com/pytorch/pytorch/issues/118790

This diff revert a small part of the code that was introduced in https://github.com/pytorch/pytorch/pull/104689

The PR above added a comment that "In case of dtype promotion, fake tensor converted into tensor"
but its not always the case that a conversion in dtype causes a fake tensor to be a tensor. 

When such conversion does not happen we get the following error
```
Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to
 a python object of type FakeTensor
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

laithsakka commented:
remove on more unexpected success

laithsakka commented:
@pytorchbot merge
